lib/switches-selectors.js

```diff
@@ -24,6 +24,9 @@ module.exports = function(switches, switchesOptions, selectors, fromEl, toEl, op
       if (this.log) {
         this.log("newEl", newEl, "oldEl", oldEl)
       }
+
+      this.state.numPendingSwitches++
```

Comment:
Incrementing `numPendingSwitches` in this loop doesn't work -- as the switch functions are executed in the same loop, multiple synchronous switches will result in `numPendingSwitches` being incremented to `1` and then immediately decremented to `0` by the new code in `onSwitch`. This then results in `afterAllSwitches` being called more than once and throwing an error on the second run as at that point the state has been reset, resulting in `this.state.options` being `null`.

It would seem safe to instead assign `numPendingSwitches` before any looping, on the very first line of the method, like so:

```js
module.exports = function(switches, switchesOptions, selectors, fromEl, toEl, options) {
  this.state.numPendingSwitches = selectors.length;

  selectors.forEach(function(selector) {
      // snip...
```

If one or more of the selectors aren't present in the switched document, the check made for each selector for `newEls.length !== oldEls.length` will throw an error and result in a full page reload being triggered.

I've tested this modification with my current project that has 3 synchronous and 1 async switches and it works for me, but obviously further testing would be required.

Reply:
Good point.

Reply:
Actually, it's more complicated than that, since it's not just one per selector, it's also times the number of `newEls`.

Reply:
Good point 😉  You're quite right, I missed that.

Might need to loop `selectors` twice then: once to build the `numPendingSwitches` correctly, and once to fire off the switches. This works for me (I cache the querySelector operation results in a map to avoid unnecessary extra queries in the second loop):

```js
var forEachEls = require("./foreach-els")

var defaultSwitches = require("./switches")

module.exports = function(switches, switchesOptions, selectors, fromEl, toEl, options) {
  var elsBySelectors = {};

  selectors.forEach(function(selector) {
    var newEls = fromEl.querySelectorAll(selector)
    var oldEls = toEl.querySelectorAll(selector)

    elsBySelectors[selector] = {
      newEls: newEls,
      oldEls: oldEls
    }

    this.state.numPendingSwitches += newEls.length
  }, this)

  selectors.forEach(function(selector) {
    var newEls = elsBySelectors[selector].newEls
    var oldEls = elsBySelectors[selector].oldEls
    if (this.log) {
      this.log("Pjax switch", selector, newEls, oldEls)
    }
    if (newEls.length !== oldEls.length) {
      // forEachEls(newEls, function(el) {
      //   this.log("newEl", el, el.outerHTML)
      // }, this)
      // forEachEls(oldEls, function(el) {
      //   this.log("oldEl", el, el.outerHTML)
      // }, this)
      throw "DOM doesn’t look the same on new loaded page: ’" + selector + "’ - new " + newEls.length + ", old " + oldEls.length
    }

    forEachEls(newEls, function(newEl, i) {
      var oldEl = oldEls[i]
      if (this.log) {
        this.log("newEl", newEl, "oldEl", oldEl)
      }

      if (switches[selector]) {
        switches[selector].bind(this)(oldEl, newEl, options, switchesOptions[selector])
      }
      else {
        defaultSwitches.outerHTML.bind(this)(oldEl, newEl, options)
      }
    }, this)
  }, this)
}
```

Alternatively, you could stick to one loop of `selectors` and build a queue of switch functions to execute after the loop is done. 

See what you think of either of those approaches...

Reply:
Maybe the queue approach is more elegant:

```js
var forEachEls = require("./foreach-els")

var defaultSwitches = require("./switches")

module.exports = function(switches, switchesOptions, selectors, fromEl, toEl, options) {
  var queue = [];

  selectors.forEach(function(selector) {
    var newEls = fromEl.querySelectorAll(selector)
    var oldEls = toEl.querySelectorAll(selector)
    if (this.log) {
      this.log("Pjax switch", selector, newEls, oldEls)
    }
    if (newEls.length !== oldEls.length) {
      // forEachEls(newEls, function(el) {
      //   this.log("newEl", el, el.outerHTML)
      // }, this)
      // forEachEls(oldEls, function(el) {
      //   this.log("oldEl", el, el.outerHTML)
      // }, this)
      throw "DOM doesn’t look the same on new loaded page: ’" + selector + "’ - new " + newEls.length + ", old " + oldEls.length
    }

    this.state.numPendingSwitches += newEls.length

    forEachEls(newEls, function(newEl, i) {
      var oldEl = oldEls[i]
      if (this.log) {
        this.log("newEl", newEl, "oldEl", oldEl)
      }

      var switchMethod = (switches[selector]) ?
        switches[selector].bind(this, oldEl, newEl, options, switchesOptions[selector]) :
        defaultSwitches.outerHTML.bind(this, oldEl, newEl, options)

      queue.push(switchMethod)
    }, this)
  }, this)

  queue.forEach(function(item) {
    item()
  })
}
```

Reply:
Even better, we don't need to increment `this.state.numPendingSwitches` manually. Instead, once all the switches are queued up, we can just do `this.state.numPendingSwitches = queue.length`.